File: codes/julia/ch11/kdv_zabusky_kruskal.jl
# kdv_zabusky_kruskal.jl
#
# Chapter 11: Fourier Pseudospectral Methods for Periodic PDEs
#
# Fourier pseudospectral solver for the KdV equation with integrating factor RK4:
#     u_t + u u_x + delta^2 u_xxx = 0 on [0, 2), periodic.
#
# Replicates the Zabusky-Kruskal (1965) experiment: a cosine initial condition
# breaks into a train of solitons, demonstrating the FPU recurrence phenomenon.
#
# Initial condition: u(x, 0) = cos(pi * x)
#
# Uses the cumulative integrating factor approach (purely imaginary linear symbol).
#
# Generates figure: kdv_zabusky_kruskal.pdf (space-time diagram)
#
# Author: Dr. Denys Dutykh
#         Mathematics Department
#         Khalifa University of Science and Technology
#         Abu Dhabi, UAE
#
# Part of the book "Computational Etudes: A Spectral Approach"
# https://github.com/dutykh/computational-etudes

using CairoMakie
using Colors
using FFTW

# -----------------------------------------------------------------------------
# Publication-quality CairoMakie configuration
# -----------------------------------------------------------------------------
set_theme!(Theme(
    fontsize = 11,
    fonts = (regular = "CMU Serif", bold = "CMU Serif Bold", italic = "CMU Serif Italic"),
    Axis = (xlabelsize = 11, ylabelsize = 11, titlesize = 12,
            xticklabelsize = 10, yticklabelsize = 10,
            spinewidth = 0.8, xtickwidth = 0.8, ytickwidth = 0.8),
    Legend = (labelsize = 10,),
))

# -----------------------------------------------------------------------------
# Configuration
# -----------------------------------------------------------------------------
const NAVY  = colorant"#142D6E"
const SKY   = colorant"#7896D2"
const CORAL = colorant"#E74C3C"
const TEAL  = colorant"#16A085"

# Output path
SCRIPT_DIR = @__DIR__
OUTPUT_DIR = joinpath(SCRIPT_DIR, "..", "..", "..", "textbook", "figures", "ch11", "julia")
mkpath(OUTPUT_DIR)

# -----------------------------------------------------------------------------
# KdV solver (Zabusky-Kruskal form)
# -----------------------------------------------------------------------------
"""
    kdv_zabusky_kruskal(; N=256, L=2.0, t_final=3.6/pi, dt=1e-4)

Solve u_t + u u_x + delta^2 u_xxx = 0 on [0, L), periodic,
with u(x, 0) = cos(pi * x) and delta^2 = 0.022^2 (Zabusky-Kruskal form).

Uses cumulative integrating factor RK4 with 2/3 dealiasing.

Returns `(x, snap_times, snapshots)` where `snapshots` is a matrix with
snapshots as rows.
"""
function kdv_zabusky_kruskal(; N::Int=256, L::Float64=2.0,
                               t_final::Float64=3.6/pi, dt::Float64=1e-4)
    delta2 = 0.022^2

    # Grid and wavenumbers
    x = L .* collect(0:N-1) ./ N
    k_base = collect(0:N-1)
    k_freq = [i <= N÷2 ? i : i - N for i in k_base]
    k = (2.0 * pi / L) .* Float64.(k_freq)

    # Linear symbol: lambda_k = -i * delta^2 * k^3
    lam = -1im .* delta2 .* k.^3

    # 2/3 dealiasing mask
    cutoff = N ÷ 3
    mask = trues(N)
    if cutoff > 0
        mask[cutoff+2:N-cutoff] .= false
    end

    # Initial condition
    u0 = cos.(pi .* x)
    v_hat = fft(u0) |> copy

    function G(t_now, v_hat_now)
        """RHS in integrating factor variables."""
        E = exp.(lam .* t_now)
        u_hat = E .* v_hat_now
        u  = real.(ifft(u_hat))
        ux = real.(ifft(1im .* k .* u_hat))
        f_nl = -u .* ux
        f_nl_hat = fft(f_nl)
        f_nl_hat[.!mask] .= 0.0
        return exp.(-lam .* t_now) .* f_nl_hat
    end

    # Time stepping
    n_steps = Int(ceil(t_final / dt))
    dt = t_final / n_steps

    n_save = 300
    save_every = max(1, n_steps ÷ n_save)
    snapshots  = [copy(u0)]
    snap_times = [0.0]

    t = 0.0
    for n in 1:n_steps
        k1 = G(t, v_hat)
        k2 = G(t + 0.5 * dt, v_hat .+ 0.5 .* dt .* k1)
        k3 = G(t + 0.5 * dt, v_hat .+ 0.5 .* dt .* k2)
        k4 = G(t + dt, v_hat .+ dt .* k3)
        v_hat .= v_hat .+ (dt / 6.0) .* (k1 .+ 2.0 .* k2 .+ 2.0 .* k3 .+ k4)
        t += dt

        if n % save_every == 0
            u_hat = exp.(lam .* t) .* v_hat
            u = real.(ifft(u_hat))
            push!(snapshots, copy(u))
            push!(snap_times, t)
        end
    end

    # Convert to matrix form
    U_mat = reduce(vcat, [u' for u in snapshots])
    return x, snap_times, U_mat
end

# -----------------------------------------------------------------------------
# Main
# -----------------------------------------------------------------------------
function main()
    N = 256
    t_final = 3.6 / pi

    x, t_save, U_save = kdv_zabusky_kruskal(N=N, t_final=t_final, dt=5e-5)

    # -------------------------------------------------------------------------
    # Figure: Space-time diagram + snapshots
    # -------------------------------------------------------------------------
    fig = Figure(size = (960, 450))

    # Left: space-time plot
    ax1 = Axis(fig[1, 1],
               xlabel = L"x",
               ylabel = L"t",
               title  = "Zabusky--Kruskal Experiment")

    hm = heatmap!(ax1, x, t_save, U_save, colormap = :RdBu, interpolate = true)
    Colorbar(fig[1, 1, Right()], hm, label = L"u(x, t)")

    # Right: snapshots at selected times
    ax2 = Axis(fig[1, 2],
               xlabel = L"x",
               ylabel = L"u(x, t)",
               title  = "Soliton Train Formation")

    t_targets = [0.0, 0.3, 0.6, 0.9, t_final]
    colors_snap = [SKY, TEAL, colorant"#2ECC71", colorant"#F39C12", NAVY]

    for (i, t_target) in enumerate(t_targets)
        idx = argmin(abs.(t_save .- t_target))
        lines!(ax2, x, U_save[idx, :], color = colors_snap[i], linewidth = 1.3,
               label = latexstring("t = $(round(t_save[idx], digits=2))"))
    end

    xlims!(ax2, 0, 2.0)
    axislegend(ax2, position = :rt, framevisible = true, labelsize = 9)
    hidespines!(ax2, :t, :r)

    save(joinpath(OUTPUT_DIR, "kdv_zabusky_kruskal.pdf"), fig)
    save(joinpath(OUTPUT_DIR, "kdv_zabusky_kruskal.png"), fig, px_per_unit = 4)
    println("Figure saved: $(joinpath(OUTPUT_DIR, "kdv_zabusky_kruskal.pdf"))")
end

main()

File: codes/matlab/ch11/kdv_zabusky_kruskal.m
%% kdv_zabusky_kruskal.m
%
% Chapter 11: Fourier Pseudospectral Methods for Periodic PDEs
%
% Fourier pseudospectral solver for the KdV equation (Zabusky-Kruskal form):
%   u_t + u u_x + delta^2 u_xxx = 0 on [0, 2), periodic.
%
% with u(x, 0) = cos(pi * x) and delta^2 = 0.022^2.
%
% Replicates the Zabusky-Kruskal (1965) experiment: a cosine initial condition
% breaks into a train of solitons.
%
% Uses integrating factor RK4 (cumulative form) with 2/3 dealiasing.
%
% Generates figure: kdv_zabusky_kruskal.pdf (space-time diagram + snapshots)
%
% Author: Dr. Denys Dutykh (Khalifa University, Abu Dhabi, UAE)
% Part of "Computational Etudes: A Spectral Approach"
% https://github.com/dutykh/computational-etudes
% Date: February 2026

clear; close all; clc;

%% Configuration
N  = 256;
L  = 2.0;
t_final = 3.6 / pi;    % Standard ZK experiment time
dt = 5e-5;
delta2 = 0.022^2;

NAVY  = [0.078 0.176 0.431];
SKY   = [0.471 0.588 0.824];
CORAL = [0.906 0.298 0.235];
TEAL  = [0.086 0.627 0.522];

% Output path
script_dir = fileparts(mfilename('fullpath'));
output_dir = fullfile(script_dir, '..', '..', '..', 'textbook', 'figures', 'ch11', 'matlab');

if ~exist(output_dir, 'dir')
    mkdir(output_dir);
end

%% Setup grid and wavenumbers
x = L * (0:N-1)' / N;

% Wavenumbers scaled for domain [0, L)
if mod(N, 2) == 0
    kn = [0:N/2-1, 0, -N/2+1:-1]';
else
    kn = [0:(N-1)/2, -(N-1)/2:-1]';
end
k = (2 * pi / L) * kn;

% Linear symbol: lambda_k = -i * delta^2 * k^3
lam = -1i * delta2 * k.^3;

% 2/3 dealiasing mask
cutoff = floor(N / 3);
mask = true(N, 1);
if cutoff > 0
    mask(cutoff+2:N-cutoff) = false;
end

%% Initial condition
u0 = cos(pi * x);
v_hat = fft(u0);

%% Time stepping: integrating factor RK4 (cumulative form)
n_steps = ceil(t_final / dt);
dt = t_final / n_steps;

n_save = 300;
save_every = max(1, floor(n_steps / n_save));
U_save = u0;
t_save = 0;

t = 0;
for n = 1:n_steps
    k1 = G_rhs(t,            v_hat,                 lam, k, mask);
    k2 = G_rhs(t + 0.5 * dt, v_hat + 0.5 * dt * k1, lam, k, mask);
    k3 = G_rhs(t + 0.5 * dt, v_hat + 0.5 * dt * k2, lam, k, mask);
    k4 = G_rhs(t + dt,       v_hat + dt * k3,         lam, k, mask);
    v_hat = v_hat + (dt / 6.0) * (k1 + 2.0 * k2 + 2.0 * k3 + k4);
    t = t + dt;

    if mod(n, save_every) == 0
        u_hat = exp(lam * t) .* v_hat;
        u = real(ifft(u_hat));
        U_save = [U_save, u]; %#ok<AGROW>
        t_save = [t_save, t]; %#ok<AGROW>
    end
end

%% Create figure
fig = figure('Units', 'inches', 'Position', [1, 1, 12, 5]);

% Left panel: space-time plot
subplot(1, 2, 1);
[X, T] = meshgrid(x, t_save);
pcolor(X, T, U_save');
shading interp;
colormap(subplot(1,2,1), redblue(256));
xlabel('$x$', 'Interpreter', 'latex', 'FontSize', 11);
ylabel('$t$', 'Interpreter', 'latex', 'FontSize', 11);
title('Zabusky--Kruskal Experiment', 'FontSize', 12);
cb = colorbar;
cb.Label.String = '$u(x, t)$';
cb.Label.Interpreter = 'latex';
cb.Label.FontSize = 11;

% Right panel: snapshots at selected times
subplot(1, 2, 2);
hold on;

t_targets = [0.0, 0.3, 0.6, 0.9, t_final];
colors = [SKY; TEAL; 0.180 0.800 0.443; 0.953 0.612 0.071; NAVY];

for i = 1:length(t_targets)
    [~, idx] = min(abs(t_save - t_targets(i)));
    plot(x, U_save(:, idx), '-', 'Color', colors(i, :), 'LineWidth', 1.3, ...
         'DisplayName', sprintf('$t = %.2f$', t_save(idx)));
end

xlabel('$x$', 'Interpreter', 'latex', 'FontSize', 11);
ylabel('$u(x, t)$', 'Interpreter', 'latex', 'FontSize', 11);
title('Soliton Train Formation', 'FontSize', 12);
lgd = legend('Interpreter', 'latex', 'FontSize', 9, 'Location', 'northeast');
lgd.BoxFace.ColorType = 'truecoloralpha';
lgd.BoxFace.ColorData = uint8([255; 255; 255; 230]);
xlim([0, L]);
grid on;
set(gca, 'GridAlpha', 0.3);
box on;

%% Save figure
exportgraphics(fig, fullfile(output_dir, 'kdv_zabusky_kruskal.pdf'), ...
               'ContentType', 'vector');
exportgraphics(fig, fullfile(output_dir, 'kdv_zabusky_kruskal.png'), ...
               'Resolution', 300);
fprintf('Figure saved: %s\n', fullfile(output_dir, 'kdv_zabusky_kruskal.pdf'));
close(fig);


%% -----------------------------------------------------------------------
%  Helper functions
%  -----------------------------------------------------------------------

function g = G_rhs(t_now, v_hat_now, lam, k, mask)
    % RHS in integrating factor variables (cumulative form)
    E = exp(lam * t_now);
    u_hat = E .* v_hat_now;
    u  = real(ifft(u_hat));
    ux = real(ifft(1i * k .* u_hat));

    % Nonlinear term: -u * u_x (ZK normalisation)
    f_nl = -u .* ux;
    f_nl_hat = fft(f_nl);
    f_nl_hat(~mask) = 0;

    g = exp(-lam * t_now) .* f_nl_hat;
end

function cmap = redblue(m)
    if nargin < 1, m = 256; end
    n = ceil(m / 2);
    r1 = linspace(0.2, 1, n)';
    g1 = linspace(0.3, 1, n)';
    b1 = linspace(0.8, 1, n)';
    r2 = linspace(1, 0.8, n)';
    g2 = linspace(1, 0.2, n)';
    b2 = linspace(1, 0.2, n)';
    cmap = [r1 g1 b1; r2(2:end,:) g2(2:end,:) b2(2:end,:)];
    if size(cmap, 1) > m
        cmap = cmap(1:m, :);
    end
end
